Mathlib/Tactic/Linter/DocModule.lean

```diff
@@ -0,0 +1,210 @@
+  Parser.testParseModule (← getEnv) "linter.style.header" (upToHere.toString ++ post)
+
+/-- `toSyntax s pattern` converts the two input strings into a `Syntax`, assuming that `pattern`
+is a substring of `s`:
+the syntax is an atom with value `pattern` whose the range is the range of `pattern` in `s`. -/
+def toSyntax (s pattern : String) (offset : String.Pos := 0) : Syntax :=
+  --let firstSubstring := ((s.splitOn pattern).getD (1 + offset) "").toSubstring
+  let beg := ((s.splitOn pattern).getD 0 "").endPos + offset
+  let fin := (((s.splitOn pattern).getD 0 "") ++ pattern).endPos + offset
+  mkAtomFrom (.ofRange ⟨beg, fin⟩) pattern
+
+-- adapted from the textbased
+/-- Return if `line` looks like a correct authors line in a copyright header. -/
+def authorsLineCorrections (line : String) (offset : String.Pos) : Array (Syntax × String) :=
+  Id.run do
+  -- We cannot reasonably validate the author names, so we look only for a few common mistakes:
+  -- the file starting wrong, double spaces, using ' and ' between names,
+  -- and ending the line with a period.
+  let mut stxs := #[]
+  if !line.startsWith "Authors: " then
+    stxs := stxs.push
+      (toSyntax line (line.take "Authors: ".length) offset,
+       s!"The authors line should begin with 'Authors: '")
+  if ((line.replace "\n  " " ").splitOn "  ").length != 1 then
+    stxs := stxs.push (toSyntax line "  " offset, s!"Double spaces are not allowed.")
+  if (line.splitOn " and ").length != 1 then
+    stxs := stxs.push (toSyntax line " and " offset, s!"Please, do not use 'and', use ',' instead.")
+  if line.endsWith "." then
+    stxs := stxs.push
+      (toSyntax line "." offset,
+       s!"Please, do not end the authors' line with a period.")
+  return stxs
+
+/-- The main function to validate the copyright string. -/
+def copyrightHeaderLinter (copyright : String) : Array (Syntax × String) := Id.run do
+  -- filter out everything after the first isolated `-/`
+  let pieces := copyright.splitOn "\n-/"
+  let copyright := (pieces.getD 0 "") ++ "\n-/"
+  let stdTxt (s : String) :=
+    s!"Malformed or missing copyright header: `{s}` should be alone on its own line."
+  let mut msgs := #[]
+  if (pieces.getD 1 "\n").take 1 != "\n" then
+    msgs := msgs.push (toSyntax copyright "-/", s!"{stdTxt "-/"}")
+  let lines := copyright.splitOn "\n"
+  let closeComment := lines.getLastD ""
+  match lines with
+  | openComment :: copyrightAuthor :: license :: authorsLines =>
+    -- The header should start and end with blank comments.
+    match openComment, closeComment with
+    | "/-", "-/" => msgs := msgs
+    | "/-", _    =>
+      msgs := msgs.push (toSyntax copyright closeComment, s!"{stdTxt "-/"}")
+    | _, _       =>
+      msgs := msgs.push (toSyntax copyright openComment, s!"{stdTxt ("/".push '-')}")
+    -- validate copyright author
+    if !copyrightAuthor.startsWith "Copyright (c) 20" then
+      msgs := msgs.push
+        (toSyntax copyright (copyrightAuthor.take 14),
```

Comment:
I would be nice to "document" the number 14 (which is the link of that string, right) --- perhaps by extract that into a variable `start` and then using `start.length` instead?

Reply:
I used the explicit length function, thanks!